sdk/storage/inc/common/retriable_stream.hpp

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "context.hpp"
+#include "http/body_stream.hpp"
+
+namespace Azure { namespace Storage {
+
+  // options used by the callback that will get a new bodyStream starting from last offset
+  struct RetryInfo
+  {
+    int64_t Offset;
+    int64_t Count;
+    std::string ETag;
+  };
+
+  struct RetriableStreamOptions
+  {
+    // configures the maximun retries to be done.
+    int64_t MaxRetryRequests;
+    // Use for testing purposes only.
+    bool DoInjectError;
+  };
+
+  // expected function signature for getting a new bodyStream upon failure.
+  // RetryInfo will be const& to avoid changes on the callback (read-only)
+  typedef std::unique_ptr<Azure::Core::Http::BodyStream> (
+      *retryFunction)(Azure::Core::Context&, RetryInfo const&);
+
+  class RetriableStream : public Azure::Core::Http::BodyStream {
+  private:
+    // initial bodyStream
+    std::unique_ptr<Azure::Core::Http::BodyStream> m_inner;
+    // Configuration for the re-triable stream
+    RetriableStreamOptions const& m_options;
```

Comment:
This should probably not be by-ref. If by-value, then the caller can get rid of their structure after calling the ctor